Support EOMONTH Transact-SQL function

pratikzode opened the pull request:
The EOMONTH function is a Transact-SQL function in SQL Server that returns the last day of the month of a specified date, with an optional offset.

1> SELECT EOMONTH ('2023-01-20')
2> go

     2023-01-31

pratikzode committed:
Support EOMONTH Transact-SQL function

The EOMONTH function is a Transact-SQL function in SQL Server that returns the last day of the month of a specified date, with an optional offset.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Updated the schedule files

Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2423,3 +2426,59 @@ pg_extension_config_remove(PG_FUNCTION_ARGS)
+
+    /* Adjust the month based on the offset */
+    month += offset;
+    // Check if the new month is greater than 0.
```
please don't mix different types of comments

pratikzode replied:
yes, will make the change in next revision

xhfanhe commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2423,3 +2426,59 @@ pg_extension_config_remove(PG_FUNCTION_ARGS)
+    {
+        ereport(ERROR,
+                (errcode(ERRCODE_DATETIME_FIELD_OVERFLOW),
+                 errmsg("Adding a value to a 'date' column caused an overflow.")));
```
Is this the same error msg as SQL server would throw?

pratikzode replied:
Yes

xhfanhe commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2423,3 +2426,59 @@ pg_extension_config_remove(PG_FUNCTION_ARGS)
+
+    /* Adjust the month based on the offset */
+    month += offset;
+    // Check if the new month is greater than 0.
```
Please modify the comment style here

pratikzode replied:
Yes, thanks for pointing that. Will do that in the next revision.

xhfanhe commented on test/JDBC/expected/sys-eomonth-vu-verify.out:
```diff
@@ -0,0 +1,322 @@
+
+
+--Edge case testing:
+--when offset is positive edge cases where yyyy can go upto 9999 after that it will throw an error.
```
Is it supposed to throw an error?

xhfanhe commented on test/JDBC/expected/sys-eomonth-vu-verify.out:
```diff
@@ -0,0 +1,322 @@
+~~ERROR (Message: Adding a value to a 'date' column caused an overflow.)~~
+
+
+--when offset is negative edge cases where yyyy can go upto 0000 after that it will throw an error.
```
Same as above, it's not throwing error here

macwoody commented on test/JDBC/expected/sys-eomonth-vu-verify.out:
```diff
@@ -0,0 +1,322 @@
+
+--Checking when separator is ‘/’ 
+--Checking valid date format for ‘YYYY/MM/DD’.
+SELECT EOMONTH ('1996/01/20',30)
```
All of these tests that tries different formats , doesn't add value to your implementation. 
As explained, they belongs to Date data type. 
Please redesign your tests to cover your own code.

pratikzode replied:
Yes, will all the test for different month which gives different last date.

pratikzode replied:
Have changed it to `LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
`

pratikzode replied:
I have added more test and removed most of the test that belongs to Date data type.
The test which I added in prepare were different month returning different last date.

macwoody commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3609,3 +3609,8 @@ IN query text)
+CREATE OR REPLACE FUNCTION sys.EOMONTH(date,int DEFAULT 0)
+RETURNS date
+AS 'babelfishpg_tsql', 'EOMONTH'
+LANGUAGE C IMMUTABLE STRICT;
```
I think this function is also parallel safe and stable .

pratikzode replied:
Yes, will make the change in the next revision.

pratikzode replied:
Will make the change

pratikzode replied:
Have changed it to `LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
`

pratikzode replied:
Have Changed from `IMMUTABLE` to `STABLE`.

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2423,3 +2426,59 @@ pg_extension_config_remove(PG_FUNCTION_ARGS)
+         * If it is, calculate the new year by adding the number of full years in the offset to the current year.
+         * The number of full years in the offset is (month - 1) / 12. 
+         */
+        year += (month - 1) / 12;
```
Could you please help explain these for calculations? 
if (month > 0 )
{
year += (month - 1) / 12;
 month = (month - 1) % 12 + 1; 
 }
 else 
{
 year += month / 12 - 1;
month = month % 12 + 12;
}

wboettge replied:
I feel several of the comments here are just restating the code. They could potentially be more helpful if they instead explained why we have the if/else here (i.e. different case with positive / negative offset). And why we need to adjust the month by one only in the positive case.

pratikzode replied:
In the expression (month - 1) / 12, the month - 1 operation is performed to account for the fact that in calendar, months are 1-indexed, meaning they start at 1 (January) and end at 12 (December).

When we add an offset to the current month, and that value is greater than 12, it means we crossed a new year. The (month - 1) / 12 operation is basically asking, "how many full cycles of 12 months are there in our offset?". So we have to think in terms of zero-based indices, hence the month - 1.

Example: If the current month is January (1) and we add an offset of 11, the result is 12 (December). In terms of full years passed, the answer is zero, even though the number is 12, because we’re still within the first year. By subtracting 1 before the division, we correctly calculate that no full year has passed.

The same logic applies to (month - 1) % 12 + 1. We subtract 1 before taking the modulus so that we're working with a zero (where January is 0 and December is 11), perform the modulus operation to find the remainder months that didn't form a full year, and then add 1 back to convert back to a 1.

pratikzode replied:
In the expression (month - 1) / 12, the month - 1 operation is performed to account for the fact that in calendar, months are 1-indexed, meaning they start at 1 (January) and end at 12 (December).

When we add an offset to the current month, and that value is greater than 12, it means we crossed a new year. The (month - 1) / 12 operation is basically asking, "how many full cycles of 12 months are there in our offset?". So we have to think in terms of zero-based indices, hence the month - 1.

Example: If the current month is January (1) and we add an offset of 11, the result is 12 (December). In terms of full years passed, the answer is zero, even though the number is 12, because we’re still within the first year. By subtracting 1 before the division, we correctly calculate that no full year has passed.

The same logic applies to (month - 1) % 12 + 1. We subtract 1 before taking the modulus so that we're working with a zero (where January is 0 and December is 11), perform the modulus operation to find the remainder months that didn't form a full year, and then add 1 back to convert back to a 1.

pratikzode replied:
Yes,I have added comments in this new revision and have explain in that.

macwoody commented on test/JDBC/expected/sys-eomonth-vu-verify.out:
```diff
@@ -0,0 +1,322 @@
+
+
+--Checking for NULL it should return NULL
+SELECT EOMONTH (NULL)
```
You may also explicitly test all combinations of NULL for your 2 arguments

pratikzode replied:
Yes, will add them in next revision.

pratikzode replied:
I have added the tests for all the combinations of NULL for your 2 arguments.

pratikzode replied:
I have added all possible combinations for NULL except 1 that is mentioned below. Because in SQL server It returns me the value for that months last date. In BBF the NULL is not getting casted to integer and so it's returning me NULL.
```
1> SELECT EOMONTH('1990-01-01',NULL)
2> go
eomonth         
------------
         NULL
```

wboettge commented on test/JDBC/expected/sys-eomonth-vu-verify.out:
```diff
@@ -0,0 +1,322 @@
+GO
+~~START~~
+date
+1996-02-29
```
I recommend adding a test for February in a leap year if that's not already included.

pratikzode replied:
yes, thanks will add that case in next revision

pratikzode replied:
Yes I have added the 2 tests for the month of feb 
--Check if its leap year. If it is a leap year and the month is feb it will return 29 otherwise 28.

pratikzode committed:
Added new test cases.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added comments in C file for explanation

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added Test cases and modified comments in c file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified sys_function.sql

Changes IMMUTABLE to STABLE

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added a new function in sql

Added a new function in sys_function.sql file to handle the case when if
second argument is NULL and also added a test case for it.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added test cases for eomonth_main

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified the code to handle NULL combinations

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Deleted the previous EOMONTH function

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added a additional check for BC date

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Alignment in C file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2423,3 +2426,98 @@ pg_extension_config_remove(PG_FUNCTION_ARGS)
+    /* If the new year is less than 1 or greater than 9999, report an error that an overflow occurred. */
+    if (year < 1 || year > 9999) 
+    {
+        if (year < -4173 || year > 5874897)
```
what does this condition mean? Please double check

pratikzode replied:
yes, have made the changes in the next revision

pratikzode committed:
Error handling for BC years

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Handling error cases for PG's date range

Signed-off-by: pratikzode <pratik.zode@gmail.com>
